content/guides/testing.mdx

```diff
@@ -0,0 +1,627 @@
+
+### Key Functions and Data Structures
+
+The key structure used is `uk_testcase` defined as:
+
+```C++
+struct uk_testcase {
+        /* The name of the test case. */
+        const char *name;
+        /* Pointer to the method  */
+        void (*func)(struct uk_testcase *self);
+        /* The number of failed assertions in this case. */
+        unsigned int failed_asserts;
+        /* The number of assertions in this case. */
+        unsigned int total_asserts;
+};
+```
+
+The macro that we're using to check conditions is `UK_TEST_ASSERT`.
+It is a wrapper over `_uk_test_do_assert`:
+
+```C++
+static inline void
+_uk_test_do_assert(struct uk_testcase *esac, bool cond, const char *fmt, ...)
+{
+        ...
+        esac->total_asserts++;
+
+        if (!cond) {
+                esac->failed_asserts++;
+        ...
+        }
+}
+```
+
+Basically, what the function does is to increment the number of failed asserts if the condition is false.
+
+We've seen that `uk_testsuite_register` is used to register tests.
+What this call boils down to is:
+
+```C++
+#define uk_test_at_initcall_prio(suite, class, prio)                   \
+        static int UK_TESTSUITE_FN(suite)(void)                        \
+        {                                                              \
+                uk_testsuite_add(&suite);                              \
+                uk_testsuite_run(&suite);                              \
+                return 0;                                              \
+        }                                                              \
+        uk_initcall_class_prio(UK_TESTSUITE_FN(suite), class, prio)
+```
+
+We can see that `uk_initcall_class_prio` registers the newly defined function as a constructor to be called at a specific time during the boot process.
+`uk_testsuite_add` simply adds the test suite to a linked listed of available test suites.
+`uk_testsuite_run` simply iterates runs all the test cases in the test suite.
+
+```C++
+int uk_testsuite_run(struct uk_testsuite *suite)
+{
+        ...
+        /* Iterate through all the registered test cases */
+        uk_testsuite_for_each_case(suite, testcase) {
+
+        /* Run the test case function testcase->func(testcase);
+
+                /* If one case fails, the whole suite fails. */
+                if (testcase->failed_asserts > 0)
+                        suite->failed_cases++;
+        }
+        ...
+}
+```
+
+## Work Items
+
+In this work session we will go over writing and running tests for Unikraft.
+We will use `uktest` and [`Google Test`](https://github.com/google/googletest).
+Make sure you are on the `usoc21` branch on the core Unikraft repo and `staging` on all others.
+`uktest` should be enabled from the Kconfig.
+
+### Support Files
+
+Session support files are available [on Google Drive](https://drive.google.com/drive/folders/1PEDtvR5W-eMGQgbJusMcDPc6hS6ghYAl?usp=sharing).
```

Comment:
We should move the items under the [exercises repo](https://github.com/unikraft-upb/guides-exercises).